LocalStorage, sessionStorage (#325)
javascript-tutorial/uk.javascript.info: 6-data-storage/02-localstorage/article.md

- `clear()` -- видалити все.
- `key(index)` -- отримати ключ на заданій позиції.
- `length` -- кількість збережених елементів.

Як ви можете бачити, це як колекція `Map` (`setItem/getItem/removeItem`), але також дозволяє доступ за індексом за допомогою `key(index)`.

Давайте подивимося, як це працює.

## Демо localStorage

Основними функціями `localStorage` є:

- Спільний доступ для всіх вкладок і вікон з одного джерела.
- Термін дії даних не закінчується. Дані залишаються після перезавантаження браузера і навіть перезавантаження ОС.

Наприклад, якщо ви запустите цей код...

```js run
localStorage.setItem('test', 1);
```

...І закриєте/відкриєте браузер або просто відкриєте ту саму сторінку в іншому вікні, то зможете отримати дані так:

```js run
alert( localStorage.getItem('test') ); // 1
```

Ми повинні бути лише в одному джерелі (домен/порт/протокол), шлях URL-адреси може бути різним.

`localStorage` використовується всіма вікнами з однаковим джерелом, тому якщо ми встановимо дані в одному вікні, зміна стане видимою в іншому.

## Доступ як до звичайного об’єкту

Ми також можемо використовувати простий об’єктний спосіб отримання/налаштування ключів, наприклад:

> ``` js run

Reviewer: ```suggestion
```js run
```

// встановити ключ
localStorage.test = 2;

// отримати ключ
alert( localStorage.test ); // 2

// видалити ключ
delete localStorage.test;
```

Це дозволено з історичних причин і в цілому працює, але зазвичай не рекомендується, оскільки:

1. Якщо ключ створений користувачем, він може бути будь-яким, наприклад, `length` або `toString`, або іншим вбудованим методом `localStorage`. У цьому випадку `getItem/setItem` працює нормально, тоді як доступ до об’єкта не вдається:
    ``` js run
    let key = 'length';
    localStorage[key] = 5; // Помилка, не вдається призначити довжину
    ```

> 2. Існує подія `storage`, вона запускається, коли ми змінюємо дані. Ця подія не відбувається при зміні даних без виклику `setItem`. Ми побачимо це пізніше в цьому розділі.

Reviewer: ```suggestion
2. Існує подія `storage`, вона запускається, коли ми змінюємо дані. Ця подія не відбувається при зміні даних без виклику `setItem`. Ми поговоримо про це пізніше в цьому розділі.
```

> ## Перебирання ключів

Reviewer: ```suggestion
## Перебір ключів
```

> Як ми бачили, методи забезпечують функцію «отримати/встановити/видалити за ключем». Але як отримати всі збережені значення чи ключі?

Reviewer: ```suggestion
Як ми бачили, методи забезпечують функцію "отримати/записати/видалити за ключем". Але як отримати всі збережені значення чи ключі?
```

На жаль, об’єкти зберігання не є ітераційними.

> Один із способів - маніпулювати ними як з масивом:

Reviewer: ```suggestion
Один із способів -- працювати з ними як з масивом:
```

}
```

> Інший спосіб — використовувати цикл `for key in localStorage`, так само, як ми робимо зі звичайними об’єктами.

Reviewer: ```suggestion
Інший спосіб -- використати цикл `for key in localStorage`, так само, як для звичайних об’єктів.
```

> Він перебирає ключі, але також виводить кілька вбудованих полів, які нам не потрібні:

Reviewer: ```suggestion
Він перебирає ключі, але також виводить кілька вбудованих властивостей, які нам не потрібні:
```

```js run

> // погана спроба

Reviewer: ```suggestion
// невдала спроба
```

for(let key in localStorage) {

>   alert(key); // показує getItem, setItem та інші вбудовані елементи

Reviewer: ```suggestion
  alert(key); // показує getItem, setItem та інші вбудовані властивості
```

}
```

...Тож нам потрібно або відфільтрувати поля з прототипу за допомогою перевірки `hasOwnProperty`:

```js run
for(let key in localStorage) {
  if (!localStorage.hasOwnProperty(key)) {
    continue; // пропускати ключі, такі як "setItem", "getItem" тощо
  }
  alert(`${key}: ${localStorage.getItem(key)}`);
}
```

> ...Або просто отримати «власні» ключі за допомогою `Object.keys`, а потім переглянути їх, якщо потрібно:

Reviewer: ```suggestion
...Або просто отримати "власні" ключі за допомогою `Object.keys`, а потім перебрати їх, якщо потрібно:
```

}
```

Останній варіант працює, оскільки `Object.keys` повертає лише ключі, які належать об’єкту, ігноруючи прототип.

## Тільки рядки

Зверніть увагу, що і ключ, і значення мають бути рядками.

> Якби був будь-який інший тип, наприклад число або об’єкт, він автоматично перетворився б на рядок:

Reviewer: ```suggestion
Якщо ми спробуємо використати будь-який інший тип, наприклад число або об’єкт, він автоматично перетвориться на рядок:
```

```js run
localStorage.user = {name: "John"};
alert(localStorage.user); // [object Object]
```

> Однак ми можемо використовувати `JSON` для зберігання об’єктів:

Reviewer: ```suggestion
Однак ми можемо використати `JSON` для зберігання об’єктів:
```

> ``` js run

Reviewer: ```suggestion
```js run
```

localStorage.user = JSON.stringify({name: "Тарас"});

// через деякий час
let user = JSON.parse( localStorage.user );
alert( user.name ); // Тарас
```

> Також можна об’єднати весь об’єкт сховища в рядки, наприклад, для цілей налагодження:

Reviewer: ```suggestion
Також можна перетворити весь об’єкт сховища на JSON рядок, наприклад, підчас налагодження коду:
```

> ``` js run

Reviewer: ```suggestion
```js run
```

// додано параметри форматування до JSON.stringify, щоб об’єкт виглядав краще
alert( JSON.stringify(localStorage, null, 2) );
```

## sessionStorage

Об’єкт `sessionStorage` використовується набагато рідше, ніж `localStorage`.

> Властивості та методи ті самі, але набагато більш обмежені:

Reviewer: ```suggestion
Властивості та методи ті самі, але можливості більш обмежені:
```

- `sessionStorage` існує лише на поточній вкладці браузера.
  - Інша вкладка з тією ж сторінкою матиме інше сховище.

>   - Але він використовується між iframes на одній вкладці (за умови, що вони походять з одного джерела).

Reviewer: ```suggestion
  - Але він використовується між iframes на одній вкладці (за умови, що це один сайт).
```

- Дані зберігаються після оновлення сторінки, але не закриття/відкриття вкладки.

Давайте подивимося на це в дії.

Запустіть цей код...

```js run
sessionStorage.setItem('test', 1);
```

> ...Тоді оновіть сторінку. Тепер ви все ще можете отримати дані:

Reviewer: ```suggestion
...Потім оновіть сторінку. Тепер ви все ще можете отримати дані:
```

> ``` js run

Reviewer: ```suggestion
```js run
```

alert( sessionStorage.getItem('test') ); // після оновлення: 1
```

...Але якщо ви відкриєте ту саму сторінку в іншій вкладці та спробуєте там ще раз, код вище поверне `null`, що означає "нічого не знайдено".

> Це саме тому, що `sessionStorage` прив’язаний не лише до джерела, а й до вкладки браузера. З цієї причини `sessionStorage` використовується не часто.

Reviewer: ```suggestion
Це саме тому, що `sessionStorage` прив’язаний не лише до сайту, а й до вкладки браузера. З цієї причини `sessionStorage` використовується не часто.
```

> ## Подія Storage

Reviewer: ```suggestion
## Подія storage
```

Коли дані оновлюються в `localStorage` або `sessionStorage`, запускається подія [storage](https://www.w3.org/TR/webstorage/#the-storage-event) із властивостями:

- `key` – ключ, який було змінено (`null`, якщо викликається `.clear()`).

> - `oldValue` – старе значення (`null`, якщо ключ додано знову).

Reviewer: ```suggestion
- `oldValue` – старе значення (`null`, якщо це новий ключ).
```

> - `newValue` – нове значення (`null`, якщо ключ видалено).

Reviewer: ```suggestion
- `newValue` – нове значення (`null`, якщо дані видалено).
```

- `url` – URL-адреса документа, де відбулося оновлення.
- `storageArea` – об’єкт `localStorage` або `sessionStorage`, де відбулося оновлення.

Важливо: подія запускається на всіх об’єктах `window`, де доступне сховище, крім того, що його викликало.

Давайте детальніше.

Уявіть, у вас є два вікна з однаковим сайтом у кожному. Таким чином, `localStorage` є спільним між ними.

```online
Ви можете відкрити цю сторінку в двох вікнах браузера, щоб перевірити наведений нижче код.
```

Якщо обидва вікна прослуховують `window.onstorage`, то кожне з них реагуватиме на оновлення, які відбулися в іншому.

```js run
// запускає оновлення, створені в тому самому сховищі з інших документів
window.onstorage = event => { // також можна використовувати window.addEventListener('storage', event => {

>   if (event.key != 'зараз') return;

Reviewer: ```suggestion
  if (event.key != 'now') return;
```

  alert(event.key + ':' + event.newValue + " в " + event.url);
};

> localStorage.setItem('зараз', ​​Date.now());

Reviewer: ```suggestion
localStorage.setItem('now', ​​Date.now());
```

```

Зверніть увагу, що подія також містить: `event.url` -- URL-адресу документа, в якому оновлено дані.

> Крім того, `event.storageArea` містить об’єкт сховища -- подія однакова для `sessionStorage` та `localStorage`, тому `event.storageArea` посилається на той, який було змінено. Ми навіть можемо захотіти щось змінити в ньому, «відреагувати» на зміни.

Reviewer: ```suggestion
Крім того, `event.storageArea` містить об’єкт сховища -- подія однакова для `sessionStorage` та `localStorage`, тому `event.storageArea` посилається на той, який було змінено. Ми навіть можемо захотіти щось змінити в ньому, "відреагувати" на зміни.
```

> **Це дозволяє різним вікнам з одного джерела обмінюватися повідомленнями.**

Reviewer: ```suggestion
**Це дозволяє різним вікнам одного сайту обмінюватися повідомленнями.**
```

> Сучасні браузери також підтримують [Broadcast channel API](mdn:/api/Broadcast_Channel_API), спеціальний API для міжвіконного зв’язку одного джерела, він більш повнофункціональний, але менш підтримується. Існують бібліотеки, які багаторазово заповнюють цей API на основі `localStorage`, що робить його доступним скрізь.

Reviewer: ```suggestion
Сучасні браузери також підтримують [Broadcast channel API](mdn:/api/Broadcast_Channel_API), спеціальний API для міжвіконного зв’язку одного джерела, він більш повнофункціональний, але має меншу підтримку серед браузерів. Існують бібліотеки-поліфіли цього API на основі `localStorage`, що робить його доступним майже скрізь.
```

> ## Висновки

Reviewer: ```suggestion
## Підсумки
```

Об’єкти веб-сховища `localStorage` та `sessionStorage` дозволяють зберігати ключ/значення в браузері.
- І `key`, і `value` мають бути рядками.
- Ліміт становить 5 Мб+, залежить від браузера.

> — Дані не мають "терміну дії", тобто не видаляються.

Reviewer: ```suggestion
- Дані не мають терміну зберіганя, тобто не видаляються.
```

> - Дані прив’язані до джерела (домен/порт/протокол).

Reviewer: ```suggestion
- Дані прив’язані до сайту (домен/порт/протокол).
```

| `localStorage` | `sessionStorage` |
|----------------|------------------|

> | Спільний для всіх вкладок і вікон з однаковим джерелом | Відображається на вкладці браузера, включаючи iframe з того самого джерела |

Reviewer: ```suggestion
| Спільний для всіх вкладок і вікон з однаковим сайтом | Відображається у вкладці браузера, включаючи iframe того самого сайту|
```

| Не видаляється при перезапуску браузера | Витримує оновлення сторінки (але не закриття вкладки) |

API:

- `setItem(key, value)` -- зберегти пару ключ/значення.
- `getItem(key)` -- отримати значення за ключем.
- `removeItem(key)` -- видалити дані за ключем.
- `clear()` -- видалити все.
- `key(index)` -- отримати ключ на заданій позиції.
- `length` -- кількість збережених елементів.
- Використовуйте `Object.keys`, щоб отримати всі ключі.

> - Ми маємо доступ до ключів як властивостей об’єкта та можемо з ними маніпулювати, у цьому випадку подія `storage` не ініціюється.

Reviewer: ```suggestion
- Ми маємо доступ до ключів як властивостей об’єкта та можемо ними маніпулювати, у цьому випадку подія `storage` не ініціюється.
```
